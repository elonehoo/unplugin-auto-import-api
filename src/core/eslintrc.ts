import type { Import } from 'unimport'
import type { ESLintGlobalsPropValue, ESLintrc } from '../types'

interface ESLintConfigs {
  globals: Record<string, ESLintGlobalsPropValue>
}

export function generateESLintConfigs(
  imports: Import[],
  eslintrc: ESLintrc,
) {
  const eslintConfigs: ESLintConfigs = { globals: {} }

  imports
    .map(i => i.as ?? i.name)
    .filter(Boolean)
    .sort()
    .forEach((name) => {
      eslintConfigs.globals[name] = eslintrc.globalsPropValue || true
    })
  const jsonBody = JSON.stringify(eslintConfigs, null, 2)
  return jsonBody
}
