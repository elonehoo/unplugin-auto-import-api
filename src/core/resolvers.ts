import { toArray } from '@antfu/utils'
import type { Addon, Import } from 'unimport'
import type { ImportExtended, ImportLegacy, Resolver, ResolverResult } from '../types'

export function normalizeImport(info: Import | ResolverResult | ImportExtended | ImportLegacy | string, name: string): ImportExtended {
  if (typeof info === 'string') {
    return {
      name: 'default',
      as: name,
      from: info,
    }
  }
  if ('path' in info) {
    return {
      from: info.path,
      as: info.name,
      name: info.importName!,
      sideEffects: info.sideEffects,
    }
  }
  return {
    name,
    as: name,
    ...info,
  }
}

export async function firstMatchedResolver(resolvers: Resolver[], fullname: string) {
  let name = fullname
  for (const resolver of resolvers) {
    if (typeof resolver === 'object' && resolver.type === 'directive') {
      if (name.startsWith('v'))
        name = name.slice(1)
      else
        continue
    }
    const resolved = await (typeof resolver === 'function' ? resolver(name) : resolver.resolve(name))
    if (resolved)
      return normalizeImport(resolved, fullname)
  }
}

export function resolversAddon(resolvers: Resolver[]): Addon {
  return {
    async matchImports(names, matched) {
      if (!resolvers.length)
        return
      const dynamic: ImportExtended[] = []
      const sideEffects: ImportExtended[] = []
      await Promise.all([...names].map(async (name) => {
        const matchedImport = matched.find(i => i.as === name)
        if (matchedImport) {
          if ('sideEffects' in matchedImport)
            sideEffects.push(...toArray((matchedImport as ImportExtended).sideEffects).map(i => normalizeImport(i, '')))

          return
        }
        const resolved = await firstMatchedResolver(resolvers, name)
        if (resolved)
          dynamic.push(resolved)

        if (resolved?.sideEffects)
          sideEffects.push(...toArray(resolved?.sideEffects).map(i => normalizeImport(i, '')))
      }))

      if (dynamic.length) {
        this.dynamicImports.push(...dynamic)
        this.invalidate()
      }

      if (dynamic.length || sideEffects.length)
        return [...matched, ...dynamic, ...sideEffects]
    },
  }
}
