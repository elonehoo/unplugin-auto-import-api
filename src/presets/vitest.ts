import type { ImportsMap } from '../types'

export default <ImportsMap>({
  vitest: [
    // suite
    'suite',
    'test',
    'describe',
    'it',
    // chai
    'chai',
    'expect',
    'assert',
    // utils
    'vitest',
    'vi',
    // hooks
    'beforeAll',
    'afterAll',
    'beforeEach',
    'afterEach',
  ],
})
