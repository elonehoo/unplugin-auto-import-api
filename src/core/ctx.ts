import { dirname, isAbsolute, relative, resolve } from 'path'
import { promises as fs } from 'fs'
import { slash, throttle, toArray } from '@antfu/utils'
import { createFilter } from '@rollup/pluginutils'
import { isPackageExists } from 'local-pkg'
import type { Import } from 'unimport'
// @ts-expect-error types
import { vueTemplateAddon } from 'unimport/addons'
import { createUnimport, scanDirExports } from 'unimport'
import MagicString from 'magic-string'
import { presets } from '../presets'
import type { ESLintrc, ImportExtended, Options } from '../types'
import { generateESLintConfigs } from './eslintrc'
import { resolversAddon } from './resolvers'

export function createContext(options: Options = {}, root = process.cwd()) {
  const imports = flattenImports(options.imports, options.presetOverriding)

  options.ignore?.forEach((name) => {
    const i = imports.find(i => i.as === name)
    if (i)
      i.disabled = true
  })

  const {
    dts: preferDTS = isPackageExists('typescript'),
  } = options

  const dirs = options.dirs?.map(dir => resolve(root, dir))

  const eslintrc: ESLintrc = options.eslintrc || {}
  eslintrc.enabled = eslintrc.enabled === undefined ? false : eslintrc.enabled
  eslintrc.filepath = eslintrc.filepath || './.eslintrc-auto-import.json'
  eslintrc.globalsPropValue = eslintrc.globalsPropValue === undefined ? true : eslintrc.globalsPropValue

  const resolvers = options.resolvers ? [options.resolvers].flat(2) : []

  const unimport = createUnimport({
    imports: imports as Import[],
    presets: [],
    injectAtEnd: true,
    addons: [
      ...(options.vueTemplate ? [vueTemplateAddon()] : []),
      resolversAddon(resolvers),
      {
        declaration(dts) {
          if (!dts.endsWith('\n'))
            dts += '\n'
          return `// Generated by 'unplugin-auto-import-api'\n${dts}`
        },
      },
    ],
  })

  const filter = createFilter(
    options.include || [/\.[jt]sx?$/, /\.vue$/, /\.vue\?vue/, /\.svelte$/],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/],
  )
  const dts = preferDTS === false
    ? false
    : preferDTS === true
      ? resolve(root, 'auto-imports.d.ts')
      : resolve(root, preferDTS)

  function generateDTS(file: string) {
    const dir = dirname(file)
    return unimport.generateTypeDeclarations({
      resolvePath: (i) => {
        if (i.from.startsWith('.') || isAbsolute(i.from)) {
          const related = slash(relative(dir, i.from).replace(/\.ts(x)?$/, ''))
          return !related.startsWith('.')
            ? `./${related}`
            : related
        }
        return i.from
      },
    })
  }

  async function generateESLint() {
    return generateESLintConfigs(await unimport.getImports(), eslintrc)
  }

  const writeConfigFilesThrottled = throttle(500, writeConfigFiles, { noLeading: false })

  async function writeFile(filePath: string, content: string) {
    await fs.mkdir(dirname(filePath), { recursive: true })
    return await fs.writeFile(filePath, content, 'utf-8')
  }

  let lastDTS: string | undefined
  let lastESLint: string | undefined
  async function writeConfigFiles() {
    const promises: any[] = []
    if (dts) {
      promises.push(
        generateDTS(dts).then((content) => {
          if (content !== lastDTS) {
            lastDTS = content
            return writeFile(dts, content)
          }
        }),
      )
    }
    if (eslintrc.enabled && eslintrc.filepath) {
      promises.push(
        generateESLint().then((content) => {
          if (content !== lastESLint) {
            lastESLint = content
            return writeFile(eslintrc.filepath!, content)
          }
        }),
      )
    }
    return Promise.all(promises)
  }

  async function scanDirs() {
    if (dirs?.length) {
      await unimport.modifyDynamicImports(async (imports) => {
        const exports = await scanDirExports(dirs, {
          filePatterns: ['*.{tsx,jsx,ts,js,mjs,cjs,mts,cts}'],
        }) as ImportExtended[]
        exports.forEach(i => i.__source = 'dir')
        return modifyDefaultExportsAlias([
          ...imports.filter((i: ImportExtended) => i.__source !== 'dir'),
          ...exports,
        ], options)
      })
    }
    writeConfigFilesThrottled()
  }

  async function transform(code: string, id: string) {
    const s = new MagicString(code)

    await unimport.injectImports(s, id)

    if (!s.hasChanged())
      return

    writeConfigFilesThrottled()

    return {
      code: s.toString(),
      map: s.generateMap({ source: id, includeContent: true }),
    }
  }

  if (!imports.length && !resolvers.length)
    console.warn('[auto-import-api] plugin installed but no imports has defined, see https://github.com/elonehoo/unplugin-auto-import-api#configurations for configurations')

  return {
    root,
    dirs,
    filter,
    scanDirs,
    writeConfigFiles,
    writeConfigFilesThrottled,
    transform,
    generateDTS,
    generateESLint,
  }
}

export function flattenImports(map: Options['imports'], overriding = false): Import[] {
  const flat: Record<string, Import> = {}
  toArray(map).forEach((definition) => {
    if (typeof definition === 'string') {
      if (!presets[definition])
        throw new Error(`[auto-import] preset ${definition} not found`)
      const preset = presets[definition]
      definition = typeof preset === 'function' ? preset() : preset
    }

    for (const mod of Object.keys(definition)) {
      for (const id of definition[mod]) {
        const meta = {
          from: mod,
        } as Import
        let name: string
        if (Array.isArray(id)) {
          name = id[1]
          meta.name = id[0]
          meta.as = id[1]
        }
        else {
          name = id
          meta.name = id
          meta.as = id
        }

        if (flat[name] && !overriding)
          throw new Error(`[auto-import] identifier ${name} already defined with ${flat[name].from}`)

        flat[name] = meta
      }
    }
  })

  return Object.values(flat)
}

function modifyDefaultExportsAlias(imports: ImportExtended[], options: Options): Import[] {
  if (options.defaultExportByFilename) {
    imports.forEach((i) => {
      if (i.name === 'default')
        i.as = i.from.split('/').pop()?.split('.')?.shift() ?? i.as
    })
  }

  return imports as Import[]
}
