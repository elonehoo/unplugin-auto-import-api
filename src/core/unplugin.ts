import { createUnplugin } from 'unplugin'
import type { Options } from '../types'
import { createContext } from './ctx'

export default createUnplugin<Options>((options) => {
  let ctx = createContext(options)
  return {
    name: 'unplugin-auto-import-api',
    enforce: 'post',
    transformInclude(id) {
      return ctx.filter(id)
    },
    async transform(code, id) {
      return ctx.transform(code, id)
    },
    async buildStart() {
      await ctx.scanDirs()
    },
    async buildEnd() {
      await ctx.writeConfigFiles()
    },
    vite: {
      async handleHotUpdate({ file }) {
        if (ctx.dirs?.some(dir => file.startsWith(dir)))
          await ctx.scanDirs()
      },
      async configResolved(config) {
        if (ctx.root !== config.root) {
          ctx = createContext(options, config.root)
          await ctx.scanDirs()
        }
      },
    },
  }
})
