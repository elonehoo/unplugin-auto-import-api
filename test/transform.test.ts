import { resolve } from 'path'
import { promises as fs } from 'fs'
import fg from 'fast-glob'
import { ElementPlusResolver } from 'unplugin-vue-components/resolvers'
import { describe, expect, it } from 'vitest'
import { createContext } from '../src/core/ctx'

describe('transform', async () => {
  const ctx = createContext({
    imports: [
      'vue',
      'pinia',
      'quasar',
      'react',
      'svelte',
      'svelte/animate',
      'svelte/easing',
      'svelte/motion',
      'svelte/store',
      'svelte/transition',
      'vitest',
      {
        'custom': [
          'customNamed',
          ['default', 'customDefault'],
          ['default', 'customDefaultAlias'],
        ],
        'ignored': [
          'ignored',
        ],
        'vue-dollar': [
          '$',
        ],
        'three.js': [
          ['*', 'THREE'],
        ],
      },
    ],
    ignore: [
      'ignored',
    ],
    resolvers: [
      (name) => {
        return name.startsWith('customResolved')
          ? `custom/resolved/${name.slice('customResolved'.length)}`
          : null
      },
      (name) => {
        return name.startsWith('customNamedResolved')
          ? {
              from: `custom/resolved/${name.slice('customNamedResolved'.length)}`,
              name: `_${name}`,
            }
          : null
      },
      ElementPlusResolver({
        importStyle: 'css',
      }),
    ],
  })

  const root = resolve(__dirname, 'fixtures')
  const files = await fg('*', {
    cwd: root,
    onlyFiles: true,
  })

  for (const file of files) {
    it(file, async () => {
      const fixture = await fs.readFile(resolve(root, file), 'utf-8')
      const pass1 = (await ctx.transform(fixture, file))?.code ?? fixture
      expect(pass1).toMatchSnapshot()
      const pass2 = (await ctx.transform(pass1, file))?.code ?? pass1
      expect(pass2).toBe(pass1)
    })
  }
})

describe('transform-vue-macro', async () => {
  const ctx = createContext({
    imports: [
      'vue/macros',
    ],
  })

  const root = resolve(__dirname, 'fixtures-vue-macro')
  const files = await fg('*', {
    cwd: root,
    onlyFiles: true,
  })

  for (const file of files) {
    it(file, async () => {
      const fixture = await fs.readFile(resolve(root, file), 'utf-8')
      const pass1 = (await ctx.transform(fixture, file))?.code ?? fixture
      expect(pass1).toMatchSnapshot()
      const pass2 = (await ctx.transform(pass1, file))?.code ?? pass1
      expect(pass2).toBe(pass1)
    })
  }
})
