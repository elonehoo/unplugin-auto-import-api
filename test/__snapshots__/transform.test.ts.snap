// Vitest Snapshot v1

exports[`transform > comments.ts 1`] = `
"/* import { ref } from 'vue' */
// import { computed } from 'vue'


import { ref, computed } from 'vue';
const a = {
  rel: \\"noreferrer\\",
  href: \\"https://github.com/antfu/vitesse\\",
  target: \\"_blank\\"
}

const foo = ref(0)
const bar = computed(() => {})

const b = { class: \\"text-sm opacity-75\\" };
// reactive
"
`;

exports[`transform > custom.js 1`] = `
"import customDefault from 'custom';
import customDefaultAlias from 'custom';
import { customNamed } from 'custom';
const a = customDefault(customNamed())
const a = customDefaultAlias()
"
`;

exports[`transform > dollar.js 1`] = `
"import { $ } from 'vue-dollar';
$('hi')
"
`;

exports[`transform > existing.jsx 1`] = `
"import ref from 'ref'
import { computed } from 'computed'


import { watch } from 'vue';
const a = ref(0)
const b = computed(0)

watch(a)
ignored(b)
"
`;

exports[`transform > existing2.jsx 1`] = `
"const { ref, computed } = Vue

const a = ref(0)
const b = computed(0)
"
`;

exports[`transform > existing3.jsx 1`] = `
"const [ ref, computed ] = Vue

const a = ref(0)
const b = computed(0)
"
`;

exports[`transform > existing4.jsx 1`] = `
"const [
  ref,
  computed
] = Vue

const a = ref(0)
const b = computed(0)
"
`;

exports[`transform > existing5.jsx 1`] = `
"import {
  ref,
  computed,
  watch as watch3,
} from 'vue'


import { watch } from 'vue';
const a = ref(0)
const b = computed(0)
const c = watch(0)
const d = watch3(0)
"
`;

exports[`transform > existing6.jsx 1`] = `
"import { computed } from 'vue';
const a = computed(() => /^(https?:\\\\/\\\\/|\\\\/\\\\/)/.test(props.to))
"
`;

exports[`transform > glob.ts 1`] = `
"import { ref } from 'vue';
const modules = import.meta.glob(\\"/src/nested/*.vue\\");
const msg = ref(\\"Global Imports\\");
"
`;

exports[`transform > import-all.ts 1`] = `
"import * as THREE from 'three.js';
const scene = new THREE.Scene();
"
`;

exports[`transform > non-target.js 1`] = `
"useNonTarget()
"
`;

exports[`transform > props.ts 1`] = `
"import { toRefs } from 'vue';
const a = z.ref(0)
console.log({
  ...toRefs(a)
})
"
`;

exports[`transform > prototype.ts 1`] = `
"const a = Object.keys(b)
const c = a.toString()

"
`;

exports[`transform > react.jsx 1`] = `
"import { useState } from 'react';
export function Component() {
  const [count, setCount] = useState()

  return <div>{ count }</div>
}
"
`;

exports[`transform > react-existing.jsx 1`] = `
"import React, { useState } from 'react';

export function Component() {
  const [count, setCount] = useState()

  return <div>{ count }</div>
}
"
`;

exports[`transform > recursive.ts 1`] = `
"export const ref = (obj: any) => {
  return ref(1)
}
"
`;

exports[`transform > resolver.ts 1`] = `
"import customResolved1 from 'custom/resolved/1';
import customResolved2 from 'custom/resolved/2';
import { _customNamedResolved2 as customNamedResolved2 } from 'custom/resolved/2';
const a = customResolved1(1)
const b = customResolved2(1)
const c = customNamedResolved2(1)
"
`;

exports[`transform > resolver-element.ts 1`] = `
"import { defineComponent } from 'vue';
export default defineComponent({
  components: {
    ElButton,
  },
  directives: {
    loading: vLoading,
  },
})

ElMessage.warning('Test')
"
`;

exports[`transform > strings.ts 1`] = `
"console.log(\\"ref(0)\\")
console.log(\`ref(\${a})\`)
console.log('ref(\\\\'z\\\\')')
console.log(\`
ref('z')
\`)
"
`;

exports[`transform > svelte.svelte 1`] = `
"import { writable } from 'svelte/store';
import { tweened } from 'svelte/motion';
import { cubicOut, quintOut } from 'svelte/easing';
import { onMount } from 'svelte';
const list = writable([1, 2, 3])
const size = tweened(1, {
  duration: 300,
  easing: cubicOut
});
onMount(() => {})

{#each $list as n (n)}
<div animate:flip={{delay: 250, duration: 250, easing: quintOut}}>
  {n}
</div>
{/each}

{#if condition}
  <div transition:fade={{amount: 10}}>
    fades in and out
  </div>
{/if}
"
`;

exports[`transform > template-literal.ts 1`] = `
"import { ref } from 'vue';
const z = \`bar-\${ref()}\`
"
`;

exports[`transform > template-ref.vue 1`] = `
"import { ref } from 'vue';
<template>
  <div ref=\\"divRef\\"></div>
</template>

<script setup>
const divRef = ref(null)
</script>
"
`;

exports[`transform > vue.js 1`] = `
"import { useQuasar, useMeta } from 'quasar';
import { ref } from 'vue';
import { defineStore, mapState } from 'pinia';
const q = useQuasar()
const meta = useMeta()
const a = ref(0)
const piniaStore = defineStore()
const piniaMapSate = mapState()
"
`;

exports[`transform-vue-macro > existing.ts 1`] = `
"import { $ } from 'vue/macros';
const not1 = /($)/.test('a'); const exist1 = $(obj); const not1 = /($)/.test('a')
"
`;

exports[`transform-vue-macro > notexisting.ts 1`] = `
"const str = '/($)/'; const test1 = /($)/.test('a')
const test2 = /[;$]/
const test3 = /($)/gim.test('a')
const test4 = /[$]/gim
const test5 = /[\\\\/$]/gim
"
`;
